rust/main/chains/hyperlane-sealevel/src/mailbox.rs

```diff
@@ -733,20 +687,178 @@ impl SealevelMailboxIndexer {
+            .get_block(dispatched_message_account.slot)
+            .await?;
+        let block_hash = decode_h256(&block.blockhash)?;
+
+        let transactions =
+            block.transactions.ok_or(HyperlaneSealevelError::NoTransactions("block which should contain message dispatch transaction does not contain any transaction".to_owned()))?;
+
+        let transaction_hashes = search_dispatched_message_transactions(
+            &self.mailbox.program_id,
+            &valid_message_storage_pda_pubkey,
+            transactions,
+        );
+
+        // We expect to see that there is only one message dispatch transaction
+        if transaction_hashes.len() > 1 {
+            Err(HyperlaneSealevelError::TooManyTransactions("Block contains more than one dispatch message transaction operating on the same dispatch message store PDA".to_owned()))?
+        }
+
+        let (transaction_index, transaction_hash) = transaction_hashes
+            .into_iter()
+            .next()
+            .ok_or(HyperlaneSealevelError::NoTransactions("block which should contain message dispatch transaction does not contain any after filtering".to_owned()))?;
+
         Ok((
             hyperlane_message.into(),
             LogMeta {
                 address: self.mailbox.program_id.to_bytes().into(),
                 block_number: dispatched_message_account.slot,
                 // TODO: get these when building out scraper support.
                 // It's inconvenient to get these :|
+                block_hash,
+                transaction_id: transaction_hash,
+                transaction_index: transaction_index as u64,
+                log_index: U256::zero(),
+            },
+        ))
+    }
+
+    fn dispatched_message_account(&self, account: &Account) -> ChainResult<Pubkey> {
+        let unique_message_pubkey = Pubkey::new(&account.data);
+        let (expected_pubkey, _bump) = Pubkey::try_find_program_address(
+            mailbox_dispatched_message_pda_seeds!(unique_message_pubkey),
+            &self.mailbox.program_id,
+        )
+        .ok_or_else(|| {
+            ChainCommunicationError::from_other_str(
+                "Could not find program address for unique message pubkey",
+            )
+        })?;
+        Ok(expected_pubkey)
+    }
+
+    async fn get_delivered_message_with_nonce(
+        &self,
+        nonce: u32,
+    ) -> ChainResult<(Indexed<H256>, LogMeta)> {
+        let discriminator = hyperlane_sealevel_mailbox::accounts::PROCESSED_MESSAGE_DISCRIMINATOR;
+        let offset = 1 + 8 + 8; // the offset to get the `message_id` field
+        let length = 32;
+        let accounts = self
+            .search_accounts_by_discriminator(&discriminator, nonce, offset, length)
+            .await?;
+
+        debug!(account_len = ?accounts.len(), "Found accounts with processed message discriminator");
+
+        let valid_message_storage_pda_pubkey = self
+            .search_and_validate_account(accounts, |account| {
```

Comment:
does this work? There's another call site as well
```suggestion
            .search_and_validate_account(accounts, self.delivered_message_account)
```

Reply:
It does not. The method takes two arguments implicitly (`&self`).